win32ss/user/user32/windows/cursoricon.c

```diff
@@ -2622,15 +2859,29 @@ HICON WINAPI CreateIconFromResourceEx(
             pbIconBits = (PBYTE)pt;
         }
 
-        if (!CURSORICON_GetCursorDataFromBMI(&cursorData, (BITMAPINFO*)pbIconBits))
+        /* Try to load BMP icon */
+        if (!CURSORICON_GetCursorDataFromBMI(&cursorData, (PBITMAPINFO)pbIconBits))
         {
-            ERR("Couldn't fill the CURSORDATA structure.\n");
-            if (ResHandle)
-                FreeResource(ResHandle);
-            return NULL;
+            /* Convert PNG raw data to BMP icon if the icon was PNG icon */
+            pbBmpIcon = CURSORICON_ConvertPngToBmpIcon(pbIconBits, cbIconBits, &BmpIconSize);
+            if (!pbBmpIcon)
+                goto end; /* Not PNG icon or failed */
+
+            /* Find icon entry from BMP icon */
+            CURSORICONFILEDIR *dir = (CURSORICONFILEDIR *)pbBmpIcon;
+            CURSORICONFILEDIRENTRY *entry = &dir->idEntries[0]; /* Only one entry */
+
+            /* A bit of preparation */
+            RtlZeroMemory(&cursorData, sizeof(cursorData));
+            cursorData.rt = LOWORD(fIcon ? RT_ICON : RT_CURSOR);
+
+            /* Can we load this BMP icon? */
+            if (!CURSORICON_GetCursorDataFromBMI(&cursorData, (PBITMAPINFO)&pbBmpIcon[entry->dwDIBOffset]))
+            {
+                ERR("Couldn't get cursor/icon data\n");
+                goto end_error;
+            }
         }
-        if (ResHandle)
-            FreeResource(ResHandle);
         isAnimated = FALSE;
```

Comment:
Could you move this assignation up to around https://github.com/reactos/reactos/pull/7709/files#diff-0bde9244bc3fdf95faed2fc129956a4ce8e298dd37f39d9ad67030e7988d6bf6R2861 (just before the first CURSORICON_GetCursorDataFromBMI call) ?
The reason is, that in your new code, you `goto end_error;` before isAnimated is set to anything, and you would end up using an uninitialized variable.

Reply:
Moved.